docs/lesson-plans/week-4.md

```diff
@@ -0,0 +1,623 @@
+address for the server. They can range from 0 to 65535, and the ones ranging from 0 to 1024 are commonly referred to as
+the [well-known port numbers](https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers#Well-known_ports). These 
+include special commonly used things such as the [File Transfer Protocol (FTP, ports 20 and 
+21)](https://en.wikipedia.org/wiki/File_Transfer_Protocol), [Secure Shell (SSH, port 22)](https://www.ssh.com/ssh/),
+[Simple Mail Transfer Protocol (SMTP, port 25)](https://en.wikipedia.org/wiki/Simple_Mail_Transfer_Protocol) (a protocol
+for sending email), DNS (port 53), (unencrypted) [Hyper Text Transfer Protocol (HTTP, port 
+80)](https://developer.mozilla.org/en-US/docs/Web/HTTP), (encrypted) [HTTP Secure (HTTPS, port 
+443)](https://developer.mozilla.org/en-US/docs/Glossary/https). When writing about ports, they can be written as either 
+"3000" or ":3000" (with the prefixed ":").
+
+HTTP/HTTPS is how websites are usually served, but 
+since those ports are often reserved by our operating system, and since we're not running this in production, it's not
+worth the effort to try and fix that. You'll see that it's common to run programs on ports outside of the well known
+port range, and ports :3000, :8000, and :8081 (what we'll be using) are quite common. Also, when testing locally,
+HTTPS is usually not used since a certificate has to be created, so just focus on using HTTP for now. We'll talk about
+HTTP more later.
+
+To access a port at a given address, you write the protocol (optional, http is assumed by default), followed by the 
+address, followed by the colon-prefixed port. So for example, accessing http on port :4000 locally would look something
+like `http://localhost:4000` or `http://127.0.0.1:4000`.
+
+#### Ngrok
+
+When you're hosting on localhost, it's often not easy for someone on a different network or even the same network to
+see the progress you've made and access your app. This is because of a bunch of firewall rules in place, and to fix that
+you will end up having to look into to port-forward your application or messing with those firewall rules. An easy way
+to get around that is to look into [Ngrok](https://ngrok.com/), which ends up forwarding traffic through one of their 
+servers for your local app and avoids having to deal with those rules. So if you want to show your friends and family,
+download it and check out their [docs](https://ngrok.com/docs). It's really easy and extremely useful!
+
+## HTTP Requests and REST
+
+### What is HTTP
+
+Hypertext Transfer Protocol (HTTP) is a protocol for requests and responses in a server-client architecture model. In
+the web browser example from earlier, the client's request to the server is an HTTP request, and the server provides an
+HTTP response in the form of actions and resources like HTML files. 
+
+HTTP provides a few useful items of information we'll be using such as headers in both requests and responses, which 
```

Comment:
"HTTP defines a few useful pieces of information in both requests and responses that we'll use. This includes..."